File: packages/mock-api/src/handlers/devices.ts
import { http, HttpResponse } from 'msw';
import type { components } from '@netvigil/shared-types';

const BASE = 'http://localhost:8000/api/v1';

const mockDevice: components['schemas']['Device'] = {
  id: '018e1234-0000-7000-8000-000000000010',
  organizationId: '018e1234-0000-7000-8000-000000000000',
  name: 'pfSense-Edge',
  vendor: 'pfsense',
  protocol: 'netflow',
  publicIp: '203.0.113.1',
  lastSeenAt: null,
  createdAt: '2024-01-01T00:00:00Z',
};

const mockList: components['schemas']['DeviceList'] = {
  items: [mockDevice],
  page: 1,
  pageSize: 25,
  total: 1,
};

export const deviceHandlers = [
  http.get(`${BASE}/devices`, () => HttpResponse.json(mockList)),
  http.post(`${BASE}/devices`, () =>
    HttpResponse.json({ ...mockDevice, sharedSecret: 'mock-secret-shown-once' }, { status: 201 }),
  ),
  http.get(`${BASE}/devices/:deviceId`, () => HttpResponse.json(mockDevice)),
  http.patch(`${BASE}/devices/:deviceId`, () => HttpResponse.json(mockDevice)),
  http.delete(`${BASE}/devices/:deviceId`, () => new HttpResponse(null, { status: 204 })),
];

File: services/api/src/netvigil_api/schemas/devices.py
from __future__ import annotations

from netvigil_api.schemas.base import CamelModel


class GeoPoint(CamelModel):
    lat: float
    lng: float


class DeviceCreate(CamelModel):
    name: str
    vendor: str
    protocol: str
    public_ip: str
    location: GeoPoint | None = None


class DeviceUpdate(CamelModel):
    name: str | None = None
    location: GeoPoint | None = None


class DeviceOut(CamelModel):
    id: str
    organization_id: str
    name: str
    vendor: str
    protocol: str
    public_ip: str
    location: GeoPoint | None
    last_seen_at: str | None
    created_at: str


class DeviceCreatedOut(DeviceOut):
    shared_secret: str


class DeviceList(CamelModel):
    items: list[DeviceOut]
    page: int
    page_size: int
    total: int
